cmd/util/cmd/find-trie-root/cmd.go

```diff
@@ -0,0 +1,232 @@
+	}
+	log.Info().Msgf("found root hash in segment %d at offset %d", segment, offset)
+
+	if len(flagOutputDir) == 0 {
+		return
+	}
+
+	err = copyWAL(flagExecutionStateDir, flagOutputDir, segment, rootHash)
+	if err != nil {
+		log.Fatal().Err(err).Msg("cannot copy WAL")
+	}
+
+	log.Info().Msgf("copied WAL to %s", flagOutputDir)
+}
+
+func parseInput(rootHashStr string) (ledger.RootHash, error) {
+	rootHashBytes, err := hex.DecodeString(rootHashStr)
+	if err != nil {
+		return ledger.RootHash(hash.DummyHash), fmt.Errorf("cannot decode root hash: %w", err)
+	}
+	rootHash, err := ledger.ToRootHash(rootHashBytes)
+	if err != nil {
+		return ledger.RootHash(hash.DummyHash), fmt.Errorf("invalid root hash: %w", err)
+	}
+	return rootHash, nil
+}
+
+func searchRootHashInSegments(
+	expectedHash ledger.RootHash,
+	dir string,
+	wantFrom, wantTo int,
+) (int, int64, error) {
+	lg := zerolog.New(os.Stderr).With().Timestamp().Logger()
+	w, err := prometheusWAL.NewSize(lg, prometheus.DefaultRegisterer, dir, wal.SegmentSize, false)
+	if err != nil {
+		return 0, 0, fmt.Errorf("cannot create WAL: %w", err)
+	}
+
+	from, to, err := prometheusWAL.Segments(dir)
+	if err != nil {
+		return 0, 0, fmt.Errorf("cannot get segments: %w", err)
+	}
+
+	if wantFrom > to {
+		return 0, 0, fmt.Errorf("from segment %d is greater than the last segment %d", wantFrom, to)
+	}
+
+	if wantTo < from {
+		return 0, 0, fmt.Errorf("to segment %d is less than the first segment %d", wantTo, from)
+	}
+
+	if wantFrom > from {
+		from = wantFrom
+	}
+
+	if wantTo < to {
+		to = wantTo
+	}
+
+	log.Info().
+		Str("dir", dir).
+		Int("from", from).
+		Int("to", to).
+		Int("want-from", wantFrom).
+		Int("want-to", wantTo).
+		Msgf("searching for trie root hash %v in segments [%d,%d]", expectedHash, wantFrom, wantTo)
+
+	sr, err := prometheusWAL.NewSegmentsRangeReader(lg, prometheusWAL.SegmentRange{
+		Dir:   w.Dir(),
+		First: from,
+		Last:  to,
+	})
+
+	if err != nil {
+		return 0, 0, fmt.Errorf("cannot create WAL segments reader: %w", err)
+	}
+
+	defer sr.Close()
+
+	reader := prometheusWAL.NewReader(sr)
+
+	for reader.Next() {
+		record := reader.Record()
+		operation, rootHash, _, err := wal.Decode(record)
+		if err != nil {
+			return 0, 0, fmt.Errorf("cannot decode LedgerWAL record: %w", err)
+		}
+
+		log.Debug().
+			Uint8("operation", uint8(operation)).
+			Str("root-hash", rootHash.String()).
+			Msgf("read LedgerWAL record")
+
+		switch operation {
+		case wal.WALUpdate:
+			if rootHash.Equals(expectedHash) {
+				log.Info().Msgf("found expected trie root hash %v", rootHash)
+				return reader.Segment(), reader.Offset(), nil
+			}
+		default:
+		}
+
+		err = reader.Err()
+		if err != nil {
+			return 0, 0, fmt.Errorf("cannot read LedgerWAL: %w", err)
+		}
+	}
+
+	return 0, 0, fmt.Errorf("finish reading all segment files from %d to %d, but not found", from, to)
+}
+
+func copyWAL(dir, outputDir string, segment int, expectedRoot ledger.RootHash) error {
```

Comment:
To help me understand:
- Is `copyWAL()` intended to create a new segment file with all records up to and including first update record with specified `rootHash`?

- If `outputDir` has existing wal files, the new segment file would be created with higher index as filename suffix.  Is this expected?

Reply:
>Is copyWAL() intended to create a new segment file with all records up to and including first update record with specified rootHash?

Yes


>If outputDir has existing wal files, the new segment file would be created with higher index as filename suffix. Is this expected?

yeah, good point. the outputDir is supposed to be different from the dir, (I can add a check to prevent that), so the generated wal is always `00000000` in a separate output folder.